examples/json_producer.py

```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+#
+# This is a simple example of the SerializingProducer using JSON.
+#
+import argparse
+from uuid import uuid4
+
+from six.moves import input
+
+from confluent_kafka import SerializingProducer
+from confluent_kafka.serialization import StringSerializer
+from confluent_kafka.schema_registry import SchemaRegistryClient
+from confluent_kafka.schema_registry.json_schema import JsonSerializer
+
+
+class User(object):
+    """
+    User record
+
+    Args:
+        name (str): User's name
+
+        favorite_number (int): User's favorite number
+
+        favorite_color (str): User's favorite color
+
+        address(str): User's address; confidential
+
+    """
+    def __init__(self, name, address, favorite_number, favorite_color):
+        self.name = name
+        self.favorite_number = favorite_number
+        self.favorite_color = favorite_color
+        # address should not be serialized, see user_to_dict()
+        self._address = address
+
+
+def user_to_dict(ctx, user):
+    """
+    Returns a dict representation of a User instance for serialization.
+
+    Args:
+        ctx (SerializationContext): Metadata pertaining to the serialization
+            operation.
+
+        user (User): User instance.
+
+    Returns:
+        dict: Dict populated with user attributes to be serialized.
+
+    """
+    # User._address must not be serialized; omit from dict
+    return dict(name=user.name,
+                favorite_number=user.favorite_number,
+                favorite_color=user.favorite_color)
+
+
+def delivery_report(err, msg):
+    """
+    Reports the failure or success of a message delivery.
+
+    Args:
+        err (KafkaError): The error that occurred on None on success.
+
+        msg (Message): The message that was produced or failed.
+
+    Note:
+        In the delivery report callback the Message.key() and Message.value()
```

Comment:
as noted previously, i think there is a good backwards compatible api here in which these are renamed `key_bytes()` and `value_bytes()` (and `key()` and `value()` retrun non-None in the bytes case). but given the experimental designation, this can wait. I just didn't want to miss an opportunity to push my opinion on this.